File: src/kernel/arch/x86-64/vm.cpp
#include "arch.h"

#include "boot.h"
#include "logger.h"

#include "utils/mem.h"

#include "vm/vm.h"

static OMA page_table_oma;

Status Arch_x86_64::vm_init() {
    constexpr auto pt_pool_pages = 512;
    void* const oma_pool = Boot::alloc(pt_pool_pages);

    if (oma_pool == Boot::alloc_fail) [[unlikely]] {
        error("Failed to allocate memory for VM page table pool");
        return KERNEL_ERROR;
    }

    const auto virt_oma_pool = VM::get_virt_dma(oma_pool);

    page_table_oma = OMA(sizeof(PageTableEntry) * page_table_size, virt_oma_pool, pt_pool_pages);
    page_table_oma.log();

    return KERNEL_OK;
}

Arch_x86_64::PageTableEntry::PageTableEntry(const uintptr_t base, const uint8_t flags) {
    present = 1;

    accessed    = 0;
    dirty       = 0;
    reserved_1  = 0;
    ignored_1   = 0;
    ignored_2   = 0;

    write_through   = 0;

    writeable       = (flags & VM::MMAP_WRITE) ? 1 : 0;
    user_access     = (flags & VM::MMAP_USER) ? 1 : 0;
    global          = (flags & VM::MMAP_GLOBAL) ? 1 : 0;
    cache_disabled  = (flags & VM::MMAP_CACHE_DISABLE) ? 1 : 0;
    exec_disabled   = (flags & VM::MMAP_EXEC) ? 0 : 1;
    size            = (flags & VM::MMAP_LARGE) ? 1 : 0;

    page_ppn = base / page_size;
}

void Arch_x86_64::PageTableEntry::prioritize_flags(const uint8_t flags) {
    writeable       |= ((flags & VM::MMAP_WRITE) != 0);
    user_access     |= ((flags & VM::MMAP_USER) != 0);
    exec_disabled   &= ((flags & VM::MMAP_EXEC) == 0);
    cache_disabled  &= ((flags & VM::MMAP_CACHE_DISABLE) != 0);
}

PageTable* PageTable::alloc() {
    PageTable* const pte = reinterpret_cast<PageTable*>(page_table_oma.alloc());

    if (pte) fill(reinterpret_cast<uint64_t*>(pte), 0ul, page_table_size);

    return pte;
}

void PageTable::free(PageTable* const page_table) {
    if (page_table == nullptr) [[unlikely]] return;

    page_table_oma.free(page_table);
}

static inline uint16_t get_pxe_idx(const uint8_t pt_idx, const uintptr_t virt_addr) {
    return (virt_addr >> ((pt_idx * 9) + 12)) & 0x1FF;
}

static inline uint64_t get_inpage_offset(const uint8_t pt_idx, const uintptr_t virt_addr) {
    return virt_addr & (~((~0xFFFul) << (pt_idx * 9)));
}

uintptr_t Arch_x86_64::get_phys(const PageTable* page_table, const uintptr_t virt_addr) {
    const PageTableEntry* pt_entry = &page_table[get_pxe_idx(3, virt_addr)];

    for (auto pt_idx = 0u; pt_idx < 4; pt_idx++) {
        if (pt_entry->present == 0) break;

        if (pt_entry->size || pt_idx == 3) {
            return pt_entry->get_base() | get_inpage_offset(3 - pt_idx, virt_addr);
        }

        pt_entry = pt_entry->get_next() + get_pxe_idx(2 - pt_idx, virt_addr);
    }

    return invalid_phys;
}

using PageTableEntry = Arch_x86_64::PageTableEntry;

namespace logging {
    static const char* prefixies[] = { "", "---|---|---", "---|---", "---" };
    static const char* size_strs[] = { "", " KB", " MB", " GB" };
    static const uint64_t size_steps[] = { 0, KB_SIZE * 4, MB_SIZE * 2, GB_SIZE };
    static const uint32_t size_units[] = { 0, 4, 2, 1 };

    static void log_pte(const PageTableEntry* pte, const uintptr_t prev_base, const uint32_t pte_idx, const uint8_t level) {
        const auto prev_idx = static_cast<uint32_t>(reinterpret_cast<uintptr_t>(pte) & 0xFFF) / 8;
        if (pte_idx - prev_idx > 1) {
            const auto last_idx = pte_idx == 511 ? pte_idx : pte_idx - 1;

            info('|', prefixies[level], 'P', level, " Entry [", prev_idx, '-', last_idx, "]: ",
                pte->get_base(), '-', prev_base + size_steps[level], ' ', (last_idx - prev_idx + 1) * size_units[level], size_strs[level]);
        }
        else {
            info('|', prefixies[level], 'P', level, " Entry [", prev_idx, "]: ", pte->get_base(), ' ', size_units, size_strs[level]);
        }
    }

    static void log_pt_helper(const PageTable* pt, const uint8_t level) {
        using PageTableEntry = Arch_x86_64::PageTableEntry;

        const PageTableEntry* pte = nullptr;
        uintptr_t prev_base = 0;

        for (auto pte_idx = 0u; pte_idx < Arch_x86_64::page_table_size; ++pte_idx) {
            const PageTableEntry& curr_pte = pt[pte_idx];

            if (curr_pte.present == 0) {
                if (pte) {
                    log_pte(pte, prev_base, pte_idx, level);
                    pte = nullptr;
                }
                continue;
            }

            if (pte && (curr_pte.size || level == 1)) {
                if (curr_pte.get_base() == prev_base + size_steps[level] &&
                    curr_pte.writeable == pte->writeable && curr_pte.exec_disabled == pte->exec_disabled) {
                    prev_base += size_steps[level];

                    if (pte_idx == 511) [[unlikely]] log_pte(pte, prev_base, pte_idx, level);
                    continue;
                }

                kassert(pte->size || level == 1);

                log_pte(pte, prev_base, pte_idx, level);

            pte_next_base:
                pte = &curr_pte;
                prev_base = curr_pte.get_base();

                if (pte_idx == 511) log_pte(pte, prev_base, pte_idx, level);
                continue;
            }
            else if (curr_pte.size || level == 1) {
                goto pte_next_base;
            }

            if (pte) {
                log_pte(pte, prev_base, pte_idx, level);
                pte = nullptr;
            }

            warn('`', prefixies[level], 'P', level, " Entry [", pte_idx, "]: ", VM::get_phys_dma(&curr_pte), " -> ", curr_pte.get_base());

            if (level > 1) log_pt_helper(curr_pte.get_next(), level - 1);
        }
    }
}

void Arch_x86_64::log_pt(const PageTable* page_table) {
    for (auto p4_idx = 0u; p4_idx < page_table_size; ++p4_idx) {
        const PageTableEntry& p4e = page_table[p4_idx];

        if (p4e.present == 0) continue;

        warn("P4 Entry [", p4_idx, "]: ", VM::get_phys_dma(&p4e));

        const PageTable* p3t = p4e.get_next();

        logging::log_pt_helper(p3t, 3);
    }
}

static constexpr auto pages_per_2_mb = ((2 * MB_SIZE) / Arch::page_size);

static inline uint8_t make_mmap_flags(uint8_t raw_flags, const uintptr_t virt, const uintptr_t phys, const uint32_t pages) {
    uint8_t result = raw_flags;

    if (raw_flags & VM::MMAP_LARGE) {
        if (pages < pages_per_2_mb ||
            (virt % (2 * MB_SIZE) != 0 || phys % (2 * MB_SIZE) != 0)
        ) {
            result ^= VM::MMAP_LARGE;
        }
    }

    return result;
}

bool Arch_x86_64::remap_large(PageTableEntry* pte, const bool is_gb_page) {
    PageTableEntry template_pte = *pte;
    template_pte.size = is_gb_page ? 1 : 0;

    PageTable* pt = PageTable::alloc();
    if (pt == PageTable::alloc_fail) return false;

    pte->page_ppn = VM::get_phys_dma(reinterpret_cast<uintptr_t>(pt)) / page_size;
    pte->size = 0;
    pte->global = 0;

    const auto pages_step = is_gb_page ? pages_per_2_mb : 1;

    for (auto i = 0u; i < page_table_size; ++i) {
        auto& entry = pt[i];

        entry = template_pte;
        template_pte.page_ppn += pages_step;
    }
    
    return true;
}

bool Arch_x86_64::early_mmap_dma() {
    PageTable* pt = VM::get_phys_dma(get_page_table());
    const auto p4_idx = get_pxe_idx(3, dma_start);
    PageTable* pt3 = reinterpret_cast<PageTable*>(Boot::alloc(1));

    {
        if (pt3 == Boot::alloc_fail) return false;

        auto& pte = pt[p4_idx];

        pte = PageTableEntry(reinterpret_cast<uintptr_t>(pt3), VM::MMAP_WRITE);
        uint64_t val = *reinterpret_cast<uint64_t*>(&pte);
    }

    PageTableEntry template_pte(static_cast<uintptr_t>(0), (VM::MMAP_GLOBAL | VM::MMAP_LARGE | VM::MMAP_WRITE));

    for (auto i = 0u; i < (dma_size / GB_SIZE); ++i) {
        pt3[i] = template_pte;
        template_pte.page_ppn += (GB_SIZE / page_size);
    }

    return true;
}

uintptr_t Arch_x86_64::mmap(
    const uintptr_t virt, const uintptr_t phys,
    const uint32_t pages, const uint8_t flags,
    PageTable* const page_table
) {
    uint8_t temp_flags = make_mmap_flags(flags, virt, phys, pages);

    PageTableEntry template_pte(phys, temp_flags);

    PageTableEntry* pt_stack[4] = { nullptr, nullptr, nullptr, nullptr };

    uint16_t pte_idx = get_pxe_idx(3, virt);
    PageTableEntry* pte = &page_table[pte_idx];

    uint32_t max_pt = 3;

    if (temp_flags & VM::MMAP_LARGE) {
        max_pt = 2;

        if (pages >= (GB_SIZE / page_size) && (virt % GB_SIZE) == 0 && (phys % GB_SIZE) == 0)
            max_pt = 1;
    }

    //debug("virt: ", virt, ": phys: ", phys, ": flags: ", flags, ": temp flags: ", temp_flags, ": max pt: ", max_pt);

    uint32_t mapped_pages = 0;

    for (auto pt_idx = 0u; pt_idx < 4;) {
        if (pt_idx < max_pt) {
            // Just lookup next entry in next page table

            if (pte->present == 0) {
                // Allocate new page table if not present
                PageTable* new_pt = PageTable::alloc();
                if (new_pt == PageTable::alloc_fail) return 0;

                *pte = template_pte;
                pte->size = 0;
                pte->global = 0;
                pte->page_ppn = VM::get_phys_dma(reinterpret_cast<uintptr_t>(new_pt)) / page_size;
            }
            else if (pte->size) {
                // Remap large page
                if (remap_large(pte, pt_idx == 1) == false) return 0;
                pte->prioritize_flags(temp_flags);
            }
            else {
                pte->prioritize_flags(temp_flags);
            }

            // Push next to the current pte on the stack
            if (pte_idx == 511) [[unlikely]] pt_stack[pt_idx] = nullptr;
            else [[likely]] pt_stack[pt_idx] = pte + 1;

            //debug("pt: ", VM::get_phys_dma(pte), " -> ", pte->get_next());

            // Go to the next pte in the next page table
            pte_idx = mapped_pages == 0 ? get_pxe_idx(2 - pt_idx, virt) : 0;
            pte = pte->get_next() + pte_idx;
            pt_idx++;
        }
        else {
            // Begin mapping
            uint32_t entries_to_map = pages - mapped_pages;
            uint32_t pages_step = 1;

            if (temp_flags & VM::MMAP_LARGE) {
                switch (max_pt) {
                case 1:
                    entries_to_map /= (GB_SIZE / page_size);
                    pages_step = (GB_SIZE / page_size);
                    break;
                case 2:
                    entries_to_map /= pages_per_2_mb;
                    pages_step = pages_per_2_mb;
                    break;
                default: kassert(false);
                }
            }

            for (; entries_to_map > 0 && pte_idx < page_table_size; ++pte_idx, --entries_to_map) {
                //debug("mmap: ", VM::get_phys_dma(pte), ": -> ", (template_pte.page_ppn + mapped_pages) * page_size, template_pte.size ? " (large)" : " (page)");

                *pte = template_pte;
                pte->page_ppn += mapped_pages;

                mapped_pages += pages_step;
                pte++;
            }

            if (entries_to_map == 0) {
                kassert(mapped_pages <= pages);

                if (mapped_pages == pages) return virt;

                kassert(temp_flags & VM::MMAP_LARGE);

                if (max_pt == 2) {
                    temp_flags ^= VM::MMAP_LARGE;
                    template_pte.size = 0;
                }

                max_pt++;
            }

            kassert(pte_idx == 512);

            while (pt_stack[pt_idx - 1] == nullptr) { kassert(pt_idx > 0); --pt_idx; }

            pte = pt_stack[--pt_idx];
            pte_idx = (reinterpret_cast<uintptr_t>(pte) & 0xFFF) / sizeof(PageTableEntry);

            kassert(pte_idx > 0);
        }
    }

    return invalid_virt;
}

void Arch_x86_64::unmap(
    const uintptr_t virt,
    const uint32_t pages,
    PageTable* const page_table
) {
     
}

void Arch_x86_64::map_ctrl(
    const uintptr_t virt,
    const uint32_t pages,
    const uint8_t flags,
    PageTable* const page_table
) {

}

File: src/kernel/arch/x86-64/vm.zig
//! # Virtual memory managment implementation

// Copyright (C) 2024 Konstantin Pigulevskiy (bagggage@github)

const std = @import("std");

const boot = @import("../../boot.zig");
const lib = @import("../../lib.zig");
const log = std.log.scoped(.@"x86-64.vm");
const regs = @import("regs.zig");
const vm = @import("../../vm.zig");

pub const page_size = 4096;

/// Linear Memory Access (LMA) region start address.
pub const lma_start = 0xFFFF800000000000;

pub const max_userspace_addr = 0x0000_8FFF_FFFF_FFFF;
pub const max_user_heap_addr = max_userspace_addr - lib.gb_size + 1;

const pages_per_2mb = (lib.mb_size * 2) / page_size;

pub const PageTable = struct {
    const Entry = packed struct {
        const Handle = struct {
            pte: *const Entry,
            pt_idx: u2
        };

        present: u1 = 0,
        writeable: u1 = 0,
        user_access: u1 = 0,
        write_through: u1 = 0,
        cache_disabled: u1 = 0,
        accessed: u1 = 0,
        dirty: u1 = 0,
        size: u1 = 0,
        global: u1 = 0,
        _ignored: u3 = 0,
        base: u28 = 0,
        _rsrvd: u12 = 0,
        _ignored2: u11 = 0,
        exec_disabled: u1 = 0,

        comptime {
            std.debug.assert(@sizeOf(@This()) == @sizeOf(u64));
            std.debug.assert(@bitSizeOf(@This()) == 64);
        }

        fn init(base: usize, flags: vm.MapFlags) @This() {
            var result: @This() = .{};

            result.present = 1;

            result.writeable = if (flags.write) 1 else 0;
            result.user_access = if (flags.user) 1 else 0;
            result.global = if (flags.global) 1 else 0;
            result.cache_disabled = if (flags.cache_disable) 1 else 0;
            result.size = if (flags.large) 1 else 0;
            result.exec_disabled = if (flags.exec) 0 else 1;

            result.base = @truncate(base / page_size);

            return result;
        }

        inline fn getBase(self: *const @This()) usize {
            return @as(usize, @intCast(self.base)) * vm.page_size;
        }

        inline fn nextPageTable(self: *const @This()) *PageTable {
            return @ptrFromInt(vm.getVirtLma(@as(usize, self.base) * page_size));
        }

        fn prioritizeFlags(self: *@This(), flags: vm.MapFlags) void {
            self.writeable |= @intFromBool(flags.write);
            self.user_access |= @intFromBool(flags.user);
            self.cache_disabled &= @intFromBool(flags.cache_disable);
            self.exec_disabled &= @intFromBool(flags.exec == false);
        }

        fn remapLarge(pte: *PageTable.Entry, is_gb_page: bool) vm.Error!void {
            var template_pte = pte.*;
            template_pte.size = if (is_gb_page) 1 else 0;

            const pt = PageTable.new() orelse return vm.Error.NoMemory;

            pte.base = @truncate(@intFromPtr(vm.getPhysLma(pt)) / page_size);
            pte.size = 0;
            pte.global = 0;

            const pages_step: u16 = if (is_gb_page) pages_per_2mb else 1;

            for (0..PageTable.len) |i| {
                pt.entries[i] = template_pte;
                template_pte.base += pages_step;
            }
        }

        fn formatHelper(self: *const Entry, writer: *std.Io.Writer, region_len: u16, level: u3) void {
            const pte_idx = (@intFromPtr(self) & 0xFFF) / @sizeOf(Entry);
            const prefix = fmt.prefixies[level];

            if (region_len > 1) {
                writer.print("{s}P{} [{}-{}]: 0x{x}->0x{x} {} {s}\n", .{
                    prefix, 4 - level, pte_idx, pte_idx + region_len - 1,
                    self.getBase(), self.getBase() + ((region_len - 1) * fmt.size_steps[level]),
                    fmt.size_units[level] * region_len, fmt.size_strs[level]
                });
            } else if (level != 3 and self.size == 0) {
                writer.print("{s}P{} [{}]: 0x{x}->0x{x}\n", .{
                    prefix, 4 - level, pte_idx,
                    vm.getPhysLma(@intFromPtr(self)), self.getBase()
                });
            } else {
                writer.print("{s}P{} [{}] -> 0x{x} {} {s}\n", .{
                    prefix, 4 - level, pte_idx, self.getBase(),
                    fmt.size_units[level], fmt.size_strs[level]
                });
            }
        }
    };

    const fmt = struct {
        const prefixies = [_][]const u8{ "", "|---", "|---|---", "|---|---|---" };
        const size_strs = [_][]const u8{ "", "GB", "MB", "KB" };
        const size_steps = [_]usize{ 0, lib.gb_size, lib.mb_size * 2, lib.kb_size * 4 };
        const size_units = [_]u8{ 0, 1, 2, 4 };
    };

    const len = 512;
    const oma_pool_pages = pages_per_2mb;

    var oma: vm.ObjectAllocator = undefined;

    entries: [len]Entry = .{ Entry{} } ** len,

    pub fn new() ?*PageTable {
        const pt = oma.alloc(PageTable) orelse return null;
        pt.* = .{};

        return pt;
    }

    pub inline fn free(self: *PageTable) void {
        oma.free(self);
    }

    pub fn translateVirtToPhys(self: *const PageTable, virt: usize) ?usize {
        const handle = self.translateVirtToPte(virt) orelse return null;
        return handle.pte.getBase() | getInpageOffset(3 - handle.pt_idx, virt);
    }

    pub fn accessPageAttributes(self: *const PageTable, virt: usize) vm.Page.Attributes {
        const pte = @constCast((self.translateVirtToPte(virt) orelse return .{}).pte);
        defer { pte.accessed = 0; pte.dirty = 0; }
        return .{
            .mapped = true,
            .writeable = pte.writeable != 0,
            .accessed = pte.accessed != 0,
            .dirty = pte.dirty != 0
        };
    }

    /// Maps a virtual memory range to a physical memory range.
    /// 
    /// - `virt`: base virtual address to which physicall region must be mapped.
    /// - `phys`: region base physical address.
    /// - `pages`: number of pages to map.
    /// - `flags`: flags to specify (see `vm.MapFlags` structure).
    /// - `page_table`: target page table.
    pub fn map(self: *PageTable, virt: usize, phys: usize, pages: u32, flags: vm.MapFlags) vm.Error!void {
        var pte_flags = correctMapFlags(flags, virt, phys, pages);
        var template_pte: Entry = .init(phys, pte_flags);

        var pt_stack: [4]?[*]Entry = .{null} ** 4;

        var pte_idx = getPxeIdx(3, virt);
        var pte: [*]Entry = self.entries[pte_idx..].ptr;

        var max_pt: u8 = 3;
        if (pte_flags.large) {
            max_pt = if (pages >= (lib.gb_size / page_size) and
                (virt % lib.gb_size) == 0 and
                (phys % lib.gb_size) == 0) 1 else 2;
        }

        var mapped_pages: u28 = 0;
        var pt_idx: u32 = 0;

        while (pt_idx < 4) {
            if (pt_idx < max_pt) {
                // Just lookup next enty in next page table
                if (pte[0].present == 0) {
                    // Allocate new page table if not present
                    const new_pt = PageTable.new() orelse return vm.Error.NoMemory;

                    pte[0] = template_pte;
                    pte[0].size = 0;
                    pte[0].global = 0;
                    pte[0].base = @truncate(@intFromPtr(vm.getPhysLma(new_pt)) / page_size);
                } else if (pte[0].size == 1) {
                    // Remap large page
                    try pte[0].remapLarge(pt_idx == 1);
                    pte[0].prioritizeFlags(pte_flags);
                } else {
                    pte[0].prioritizeFlags(pte_flags);
                }

                // Push next to the current pte on the stack
                if (pte_idx == 511) {
                    pt_stack[pt_idx] = null;
                } else {
                    pt_stack[pt_idx] = pte + 1;
                }

                // Go to the next pte in the next page table
                pte_idx = if (mapped_pages == 0) getPxeIdx(@truncate(2 - pt_idx), virt) else 0;
                pte = pte[0].nextPageTable().entries[pte_idx..].ptr;
                pt_idx += 1;
            } else {
                // Begin mapping
                var entries_to_map = pages - mapped_pages;
                var pages_step: u28 = 1;

                if (pte_flags.large) {
                    switch (max_pt) {
                        1 => {
                            pages_step = lib.gb_size / page_size;
                            entries_to_map /= pages_step;
                        },
                        2 => {
                            pages_step = pages_per_2mb;
                            entries_to_map /= pages_step;
                        },
                        else => unreachable,
                    }
                }

                while (entries_to_map > 0 and pte_idx < PageTable.len) : ({
                    pte_idx += 1;
                    entries_to_map -= 1;
                }) {
                    pte[0] = template_pte;
                    pte[0].base += mapped_pages;

                    mapped_pages += pages_step;
                    pte += 1;
                }

                if (entries_to_map == 0) {
                    std.debug.assert(mapped_pages <= pages);

                    if (mapped_pages == pages) return;

                    std.debug.assert(pte_flags.large);

                    if (max_pt == 2) {
                        pte_flags.large = false;
                        template_pte.size = 0;
                    }

                    max_pt += 1;
                }

                std.debug.assert(pte_idx == PageTable.len);

                while (pt_stack[pt_idx - 1] == null) {
                    pt_idx -= 1;
                }

                pt_idx -= 1;
                pte = pt_stack[pt_idx].?;
                pte_idx = @truncate((@intFromPtr(pte) & 0xFFF) / @sizeOf(Entry));

                std.debug.assert(pte_idx > 0);
            }
        }

        unreachable;
    }

    pub fn unmap(self: *PageTable, virt: usize, pages: u32) void {
        var pt_stack: [4]?[*]Entry = .{null} ** 4;

        var pte_idx = getPxeIdx(3, virt);
        var pte: [*]Entry = self.entries[pte_idx..].ptr;

        var mapped_pages: u28 = 0;
        var pt_idx: u32 = 0;

        while (pt_idx < 4) {
            if (pt_idx < 3) {
                // Just lookup next entry in next page table
                if (pte[0].present == 0) {
                    // Page table not present, nothing to unmap
                    return;
                } else if (pte[0].size != 0) {
                    // Remap large page
                    pte[0].remapLarge(pt_idx == 1) catch {
                        // TODO: fix it!
                        return;
                    };
                }

                // Push next to the current pte on the stack
                if (pte_idx == 511) {
                    pt_stack[pt_idx] = null;
                } else {
                    pt_stack[pt_idx] = pte + 1;
                }

                // Go to the next pte in the next page table
                pte_idx = if (mapped_pages == 0) getPxeIdx(@truncate(2 - pt_idx), virt) else 0;
                pte = pte[0].nextPageTable().entries[pte_idx..].ptr;
                pt_idx += 1;
            } else {
                // Begin unmapping
                var entries_to_unmap = pages - mapped_pages;
                var pages_step: u28 = 1;

                // Check if this is a large page
                if (pte[0].size == 1) {
                    // Determine if this is a 1GB or 2MB page
                    const is_1gb = (pt_idx == 1);
                    pages_step = if (is_1gb)
                        (lib.gb_size / page_size)
                    else
                        pages_per_2mb;

                    entries_to_unmap = 1; // Large page covers all remaining entries
                }

                while (entries_to_unmap > 0 and pte_idx < PageTable.len) : ({
                    pte_idx += 1;
                    entries_to_unmap -= 1;
                }) {
                    pte[0].present = 0;
                    mapped_pages += pages_step;
                    pte += 1;
                }

                if (entries_to_unmap == 0) {
                    if (mapped_pages == pages) return;
                }

                std.debug.assert(pte_idx == PageTable.len);

                // Go back up the stack
                while (pt_idx > 0 and pt_stack[pt_idx - 1] == null) {
                    pt_idx -= 1;
                }

                if (pt_idx == 0) return;

                pt_idx -= 1;
                pte = pt_stack[pt_idx].?;
                pte_idx = @truncate((@intFromPtr(pte) & 0xFFF) / @sizeOf(Entry));

                std.debug.assert(pte_idx > 0);
            }
        }
    }

    pub inline fn format(writer: *std.Io.Writer, self: *const PageTable) void {
        self.formatLevel(writer, 0);
    }

    fn formatLevel(self: *const PageTable, writer: *std.Io.Writer, level: u2) void {
        var pte_idx: u16 = 0;

        while (pte_idx < PageTable.len) : (pte_idx += 1) {
            const pte = &self.entries[pte_idx];
            if (pte.present == 0) continue;

            if (pte.size == 1 or level == 3) {
                const max_len = PageTable.len - pte_idx;
                var region: u16 = 1;
                for (1..max_len) |i| {
                    const idx = pte_idx + i;

                    if (self.entries[idx].base != pte.base + ((fmt.size_steps[level] / page_size) * i) or
                        self.entries[idx].present == 0 or i == max_len - 1)
                    {
                        region = @truncate(i);
                        break;
                    }
                }

                pte.formatHelper(writer, region, level);
                pte_idx += region - 1;
            } else {
                pte.formatHelper(writer, 1, level);
                pte.nextPageTable().formatLevel(writer, level + 1);
            }
        }
    }

    fn correctMapFlags(flags: vm.MapFlags, virt: usize, phys: usize, pages: u32) vm.MapFlags {
        var result = flags;
        if (flags.large) {
            if (pages < pages_per_2mb or
                (virt % (2 * lib.mb_size) != 0 or
                    phys % (2 * lib.mb_size) != 0)) result.large = false;
        }

        return result;
    }

    fn translateVirtToPte(self: *const PageTable, virt: usize) ?Entry.Handle {
        var pte = &self.entries[getPxeIdx(3, virt)];
        for (0..4) |pt_idx| {
            if (pte.present == 0) break;
            if (pte.size == 1 or pt_idx == 3) return .{ .pte = pte, .pt_idx = @truncate(pt_idx) };

            pte = &pte.nextPageTable().entries[getPxeIdx(@truncate(2 - pt_idx), virt)];
        }

        return null;
    }

    inline fn getPxeIdx(pt_idx: u8, virt: usize) u16 {
        return @truncate((virt >> @truncate((pt_idx * 9) + 12)) & 0x1FF);
    }

    inline fn getInpageOffset(pt_idx: u8, virt: usize) u12 {
        return @truncate(virt & ~(~@as(usize, 0xFFF) << @truncate(pt_idx * 9)));
    }
};

var lma_end: usize = 0;
var heap_start: usize = 0;

pub fn preinit() void {
    earlyMapLma();
    boot.switchToLma();
}

pub fn init() vm.Error!void {
    const rank = std.math.log2(PageTable.oma_pool_pages);
    const oma_pool = vm.PageAllocator.alloc(rank) orelse return vm.Error.NoMemory;

    PageTable.oma = try .initRaw(@sizeOf(PageTable), oma_pool, PageTable.oma_pool_pages);
}

pub inline fn lmaEnd() usize {
    return lma_end;
}

pub inline fn heapStart() usize {
    return heap_start;
}

pub inline fn isUserVirtAddr(virt: usize) bool {
    return (virt & 0xFFFF_0000_0000_0000) == 0;
}

pub inline fn getPageTable() *PageTable {
    const pt: *PageTable = @ptrFromInt(regs.getCr3() & ~@as(usize, 0xFFF));
    return vm.getVirtLma(pt);
}

pub inline fn setPageTable(pt: *const PageTable) void {
    const pt_phys = vm.getPhysLma(@intFromPtr(pt));
    const cr3 = pt_phys | (regs.getCr3() & @as(u64, 0xFFF));

    regs.setCr3(cr3);
}

pub fn copyKernelMappings(source: *const PageTable, dest: *PageTable) void {
    const lma_pte_idx = comptime PageTable.getPxeIdx(3, lma_start);
    const heap_pte_idx = PageTable.getPxeIdx(3, heap_start);
    const kernel_pte_idx = PageTable.getPxeIdx(3, @intFromPtr(&init));

    dest.entries[lma_pte_idx] = source.entries[lma_pte_idx];
    dest.entries[heap_pte_idx] = source.entries[heap_pte_idx];
    dest.entries[kernel_pte_idx] = source.entries[kernel_pte_idx];
}

fn earlyMapLma() void {
    const lma_size = calcLmaSize();

    lma_end = lma_start + lma_size;
    heap_start = lma_end + lib.gb_size;

    const pt = vm.getPhysLma(getPageTable());
    const p4_idx = comptime PageTable.getPxeIdx(3, lma_start);

    const pt3: *PageTable = @ptrFromInt(boot.alloc(1).?);
    @memset(@as(*[PageTable.len]u64, @ptrCast(pt3))[0..PageTable.len], 0);

    const pte: *PageTable.Entry = &pt.entries[p4_idx];
    pte.* = .init(@intFromPtr(pt3), vm.MapFlags{ .write = true });

    var template_pte: PageTable.Entry = .init(0, vm.MapFlags{ .write = true, .global = true, .large = true });
    const len = lma_size / lib.gb_size;
    const gb_pages = lib.gb_size / vm.page_size;

    for (pt3.entries[0..len]) |*entry| {
        entry.* = template_pte;
        template_pte.base += gb_pages;
    }
}

fn calcLmaSize() usize {
    // TODO: implement this to ensure that
    // all physical memory is accessible via LMA.

    // Currently just return 256 GB.
    return 256 * lib.gb_size;
}
